main.go

```diff
@@ -0,0 +1,225 @@
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sort"
+	"strings"
+
+	"github.com/Masterminds/semver/v3"
+)
+
+type CincinnatiGraph struct {
+	Nodes []CincinnatiNode `json:"nodes"`
+}
+
+type CincinnatiNode struct {
+	Version  *semver.Version   `json:"version"`
+	Payload  string            `json:"payload"`
+	Metadata map[string]string `json:"metadata"`
+}
+
+type Release struct {
+	Version  *semver.Version
+	Channel  string
+	Arch     string
+	Payload  string
+	Metadata map[string]string
+}
+
+type VersionReleases map[string]Release
+
+type ReleasesByChannel map[string]VersionReleases
+
+// fetchGraph fetches the upgrade graph for a given channel and architecture.
+func fetchGraph(client *http.Client, channel, arch string) (*CincinnatiGraph, error) {
+	u, err := url.Parse("https://api.openshift.com/api/upgrades_info/graph")
+	if err != nil {
+		return nil, err
+	}
+	queryParams := u.Query()
+	queryParams.Add("channel", channel)
+	queryParams.Add("arch", arch)
+	u.RawQuery = queryParams.Encode()
+
+	resp, err := client.Get(u.String())
+	if err != nil {
+		return nil, fmt.Errorf("error fetching data from %s: %w", u, err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("error: status %d when fetching data from %s", resp.StatusCode, u)
+	}
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("error reading response from %s: %w", u, err)
+	}
+	var graph CincinnatiGraph
+	if err = json.Unmarshal(body, &graph); err != nil {
+		return nil, fmt.Errorf("error parsing JSON from %s: %w", u, err)
+	}
+	return &graph, nil
+}
```

Comment:
why are you not using the Cincinnati client directly?
https://github.com/openshift/cluster-version-operator/blob/main/pkg/cincinnati/cincinnati.go

Reply:
dep on an external repo, besides we just fetch the graph

Reply:
there's already plenty of external repositories from openshift in CS that they depend on
![image](https://github.com/user-attachments/assets/8e9e34fd-e508-4487-8046-aa38c2235988)

Reply:
if there is a strong desire i think we could use but:

- it doesn't look like an official cincinnati client, just a client the CVO uses.
- the client requires `version` (the version number of the currently installed cluster) and `id` (the unique identifier (UUID v4) of the cluster) which we don't have.
- the client returns strong types which i think we also don't need.

WDYT ?

Reply:
@wking my assumption is that the "id" is just an identifier to understand the origin of the request, rather than a lookup in some database. So in our case that would just be "ARO-HCP ClustersService"? 

> the client returns strong types which i think we also don't need.

you don't want to use the official Openshift/API types? 

> if there is a strong desire

I just want to reduce the maintenance burden of the ARO folks. I think they would be happy when someone takes care of the CVO integration that knows CVO.

Reply:
the `version` param is used in the code - https://github.com/openshift/cluster-version-operator/blob/main/pkg/cincinnati/cincinnati.go#L161

Reply:
It's all obviously public, Apache-2.0 code, but yeah, the library is currently mostly CVO-internal.  I'm not opposed to moving some/all of it into [library-go][1] and being a bit more intentional about the API we decide to expose, if other folks want to share.  As you point out, the current CVO code is geared towards "I'm a cluster asking about my own update advice" and would need adjustments to support "I'm a generic client interested in Cincinnati Update Service opinions, and not a cluster myself".  But I also think the Go wrapping is straightforward and stable enough that I don't have a problem with consuming teams rolling their own Go wrapper without collecting around a shared library.

[1]: https://github.com/openshift/library-go

Reply:
I would prefer to own a go wrapper for now.